### ISSUE TITLE
DiagnosticBuilder::span_label (sometimes?) does nothing

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#71857]
Recently, I was constructing a diagnostic with code like this:
```rust
            let mut diag = self.struct_warn("skipping const checks");
            for &(span, feature_gate) in unleashed_features.iter() {
                if let Some(feature_gate) = feature_gate {
                    diag.span_label(span, format!("skipping check for `{}` feature", feature_gate));
                } else {
                    diag.span_label(span, "skipping check that does not even have a feature gate");
                }
            }
            diag.emit();
```
I expected this to print a warning with a bunch of labels pointing to the given spans attached to it. But instead, the labels are just entirely ignored.

I am now using `span_help` instead, but that's a gross hack as this is not a "help" information. (This warning is for permanently unstable nightly-only testing flags, so it's not a big deal.) Unfortunately the `span_label` doc comment is also not very helpful, talking about the internal representation of the diagnostic builder ("multispan" etc) and about what happens when I do *not* call this method (?!?) but not about why adding a label might just have no effect at all.

Cc @rust-lang/wg-diagnostics which hopefully has people that actually understand this API, unlike me. :)

**Comment by estebank** [comment]
Did span_help show a subdiagnostic in the output or just the `help`? If the span isn't displayable (points at code that rustc can't find or it's a DUMMY_SP) then span_label does nothing because it assumes that the label without the context isn't understandable, while note and help assume that the text might be enough.

**Comment by RalfJung** [comment]
> If the span isn't displayable (points at code that rustc can't find or it's a DUMMY_SP) then span_label does nothing because it assumes that the label without the context isn't understandable, while note and help assume that the text might be enough.

That kind of detail would be really good to have state in the doc comments of the involved functions. :D

Final output looks like this:
```
warning: skipping const checks
   |
help: skipping check that does not even have a feature gate
  --> $DIR/const_refers_to_static2.rs:14:18
   |
LL |     unsafe { &*(&FOO as *const _ as *const usize) }
   |                  ^^^
help: skipping check for `const_raw_ptr_deref` feature
  --> $DIR/const_refers_to_static2.rs:14:14
   |
LL |     unsafe { &*(&FOO as *const _ as *const usize) }
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: skipping check that does not even have a feature gate
  --> $DIR/const_refers_to_static2.rs:22:6
   |
LL |     &FOO
   |      ^^^
```
Indeed the "root warning" has no span we could attach it to.
